### PR TITLE
fix: resolve with cwd when context is empty string

### LINKED ISSUE #2067: [Bug]: when virtual module includes a relative path import, `transform(code, id)` hook gives `id` as a relative path
### Reproduction link or steps

reproduction: https://github.com/hi-ogawa/reproductions/tree/main/rolldown-transform-id-not-absolute

I'm not sure if this is a bug, but I just found a minor difference from rollup's behavior while checking rolldown compatibility of module federation plugin https://github.com/MadaraUchiha-314/rollup-plugin-module-federation/.

_Here pasting readme from the reproduction above_

When a virtual module includes a relative import e.g.

```ts
// virtual:entry
import * as lib from "./src/index.js";
console.log(lib);
```

Rollup provides a resolved absolute path in `transform(code, id)` hook:

```sh
$ pnpm build-rollup

[debug:transform] {
  id: '/home/hiroshi/code/personal/reproductions/rolldown-transform-id-not-absolute/src/index.js',
  code: 'console.log("test");\n'
}
```

On the other hand, Rolldown provides a relative path:

```sh
$ pnpm build-rolldown

[debug:transform] { id: 'src/index.js', code: 'console.log("test");\n' }
```


### What is expected?

Should work like rollup? or maybe reject such relative import from virtual module?
I would expect plugin can usually resolve relative paths on its own for virtual module.

### What is actually happening?

`id` in `transform(code, id)` hook is not resolved

### System Info

```Shell
System:
    OS: Linux 6.10 Arch Linux
    CPU: (16) x64 12th Gen Intel(R) Core(TM) i7-12650H
    Memory: 20.28 GB / 31.06 GB
    Container: Yes
    Shell: 5.2.32 - /usr/bin/bash
  Binaries:
    Node: 20.16.0 - ~/.volta/tools/image/node/20.16.0/bin/node
    npm: 10.8.1 - ~/.volta/tools/image/node/20.16.0/bin/npm
    pnpm: 9.8.0 - ~/.volta/bin/pnpm
  npmPackages:
    rolldown: 0.12.2-snapshot-8ea9f31-20240826003035 => 0.12.2-snapshot-8ea9f31-20240826003035
```


### Any additional comments?

I also tried wasi version on stackblitz https://stackblitz.com/edit/github-n7gv1v?file=dist%2Frolldown%2Fmain.js and this version cannot resolve it, so it's treated as external:

```
 npm run build-rolldown

> build-rolldown
> rolldown -c rolldown.config.mjs

(node:28) ExperimentalWarning: WASI is an experimental feature and might change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
(node:28) ExperimentalWarning: WASI is an experimental feature and might change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
(node:28) ExperimentalWarning: WASI is an experimental feature and might change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
[UNRESOLVED_IMPORT] Warning: "./src/index.js" is imported by "\0virtual:entry", but could not be resolved – treating it as an external dependency.

<DIR>/main.js  chunk │ size: 0.06 kB                                                                                   
                                                                                                                       
✔ Finished in 495.23 ms                                                                                               
```

### LINKED ISSUE #2269: [Bug]: node module resolution not working from virtual modules
### Reproduction link or steps

- stackblitz https://stackblitz.com/edit/github-mcv3k1?file=rolldown.config.mjs
- comparison with rollup https://github.com/hi-ogawa/reproductions/tree/main/rolldown-virtual-resolve-node-modules

When bundling a following code as an actual js file `actual.js` and virtual file `virtual.js`:

```js
import slash from "slash";
console.log(slash);
```

node module fails for the virtual module:

```
[UNRESOLVED_IMPORT] Warning: "slash" is imported by "\0virtual.js", but could not be resolved – treating it as an external dependency.

<DIR>/virtual.js  chunk │ size: 0.10 kB
<DIR>/actual.js   chunk │ size: 0.31 kB
```

### What is expected?

node module resolution works

### What is actually happening?

node module resolution not working

### System Info

```Shell
(stackblitz)

  System:
    OS: Linux 5.0 undefined
    CPU: (8) x64 Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
    Memory: 0 Bytes / 0 Bytes
    Shell: 1.0 - /bin/jsh
  Binaries:
    Node: 18.20.3 - /usr/local/bin/node
    Yarn: 1.22.19 - /usr/local/bin/yarn
    npm: 10.2.3 - /usr/local/bin/npm
    pnpm: 8.15.6 - /usr/local/bin/pnpm
  npmPackages:
    rolldown: latest => 0.13.2
```

### Any additional comments?

I found this when testing rollup module federation plugin https://github.com/hi-ogawa/rollup-plugin-module-federation/pull/1 as it relies on virtual module which imports some node modules.

Maybe this is similar to https://github.com/rolldown/rolldown/issues/2067 and could it be solved by the proposed PR? https://github.com/rolldown/rolldown/pull/2080

### PR DESCRIPTION
**Pull request by inottn**
### Description

fix #2067

### CODE REVIEW

File: packages/rolldown/src/rolldown-binding.wasi-browser.js [hunk 1]
```diff
@@ -70,53 +70,65 @@ function __napi_rs_initialize_modules(__napiInstance) {
   __napiInstance.exports['__napi_register__transform_9']?.()
   __napiInstance.exports['__napi_register__Bundler_struct_0']?.()
   __napiInstance.exports['__napi_register__Bundler_impl_5']?.()
-  __napiInstance.exports['__napi_register__BindingInputItem_struct_6']?.()
```
Comment (7086cmd) [inline packages/rolldown/src/rolldown-binding.wasi-browser.js]: Why there are changes in napi binding?

Comment (inottn) [reply packages/rolldown/src/rolldown-binding.wasi-browser.js]: I'm also curious that when I run the `just test` command, this file automatically changes.

**Comment by hyf0** [comment]
Could you add some description to explain the issue and the fix?

**Comment by inottn** [comment]
> Could you add some description to explain the issue and the fix?

This fix aligns with Rollup's implementation. In this scenario, when the parent directory returns an empty string, path.resolve('', source) is equivalent to path.resolve(source), which resolves source as an absolute path relative to the cwd.

**Comment by hi-ogawa** [comment]
I had an issue with `node_modules` resolution also not working inside virtual module https://github.com/rolldown/rolldown/issues/2269. Would this PR also fix this case?

**Comment by hyf0** [comment]
I'll take a deep look today.

**Comment by underfin** [comment]
I debugger the rollup code, rollup using `resolve(dirname(importer), source)` to resolve it, here has some result from node
```
> require('path').dirname('c') // here returned '.'
'.'
> require('path').resolve('c')
'/Users/likui/Github/rollup/c'
> require('path').resolve('.', 'c')
'/Users/likui/Github/rollup/c'
```

The pr using empty string to check it, i think it is enough to aligin the rollup behavior.

File: crates/rolldown_resolver/src/resolver.rs [hunk 2]
```diff
@@ -157,26 +157,19 @@ impl<F: FileSystem + Default> Resolver<F> {
     };
     let resolution = if let Some(importer) = importer {
       let context = importer.parent().expect("Should have a parent dir");
-      selected_resolver.resolve(context, specifier)
+
+      if context.to_str().map_or(true, str::is_empty) {
```
Comment (underfin) [inline crates/rolldown_resolver/src/resolver.rs]: Here is better to add some notes for catching virtual files.

**Comment by underfin** [comment]
> I had an issue with `node_modules` resolution also not working inside virtual module #2269. Would this PR also fix this case?

It should be fixed https://github.com/rolldown/rolldown/issues/2269

File: crates/rolldown_resolver/src/resolver.rs (hunk 2 above)
Comment (inottn) [reply crates/rolldown_resolver/src/resolver.rs]: done

File: packages/rolldown/tests/fixtures/builtin-plugin/alias/at-alias-prefix/_config.ts [hunk 3]
```diff
@@ -2,6 +2,7 @@ import { aliasPlugin } from 'rolldown/experimental'
 import { defineTest } from '@tests'
 
 export default defineTest({
+  skip: true,
```
Comment (hyf0) [inline packages/rolldown/tests/fixtures/builtin-plugin/alias/at-alias-prefix/_config.ts:5]: This test failed in this PR, which considered expected.

```
aliasPlugin({
        entries: [
          { find: '@src', replacement: 'src' },
          { find: '@utils', replacement: './utils' },
        ],
      }),
```

would transform

```
import '@src/index.js'
```

to 

```
import 'src/index.js'
```

`import 'src/index.js'` is not handleable with default configuration of rolldown.

@IWANABETHATGUY cc

**Review by hyf0 — APPROVED** [review]
Thanks

File: packages/rolldown/tests/fixtures/builtin-plugin/alias/at-alias-prefix/_config.ts (hunk 3 above)
Comment (hyf0) [reply packages/rolldown/tests/fixtures/builtin-plugin/alias/at-alias-prefix/_config.ts:5]: I'll merge this PR first. If we should make `{ find: '@src', replacement: 'src' },` work by default in rolldown, let's have a discussion then.

**Pull request merged**